src/main/vcpf4/usbd_cdc_vcp.c

```diff
@@ -193,12 +192,17 @@ uint32_t CDC_Receive_DATA(uint8_t* recvBuf, uint32_t len)
 {
     uint32_t count = 0;
 
-    while (usbAvailable() && count < len) {
-        recvBuf[count] = usbData.buffer[usbData.bufferOutPosition];
-        usbData.bufferOutPosition = (usbData.bufferOutPosition + 1) % USB_RX_BUFSIZE;
+    while (APP_Tx_ptr_out != APP_Tx_ptr_in && count < len) {
+        recvBuf[count] = APP_Tx_Buffer[APP_Tx_ptr_out];
+        APP_Tx_ptr_out = (APP_Tx_ptr_out + 1) % USB_RX_BUFSIZE;
         count++;
         receiveLength--;
     }
+
+    if (!receiveLength) {
+        receiveLength = APP_Tx_ptr_out != APP_Tx_ptr_in;
```

Comment:
It is used in the function serialRxBytesWaiting (all over the place) - which in turn is referencing the serial vTable pointer to the usbVcpAvailable function. If you search, it is invariably always used as a bool - even though it returns an uint32_t which is the count.

Rather than returning 1, I could calculate the difference between ptr_in and ptr_out, but given the code all just uses it as an "are there more than 0 bytes available check" there really is no point in doing the calculation (at least at the moment).

For the circular buffer override check it would be easier to merely add a condition in the for loop - and then return fail within the loop. See below: